# Investigation of borrowers' reliability

### Description of the project

The customer is the credit department of a bank. It is necessary to determine whether the client's family status and number of children affect the timely repayment of the loan. The input data from the bank is the statistics on the clients' solvency.

The results of the research will be taken into account when constructing a credit scoring model - a special system that assesses the ability of a potential borrower to repay the loan to the bank.

### Description of Data

* children — number of children in the family
* days_employed — total work experience in days
* dob_years — age of the client in years
* education — level of education of the client
* education_id — identifier of the level of education
* family_status — family status
* family_status_id — identifier of the family status
* gender — gender of the client
* income_type — type of employment
* debt — debt for credit repayment
* total_income — monthly income
* purpose — reason for obtaining credit

### Plan for conducting the research:

1. Overview of the data received;
2. Preparation of the data for analysis and hypothesis testing;
3. Data analysis and hypothesis testing;
4. Research results.

## Data overview

In [131]:
import pandas as pd

In [132]:
url = 'https://code.s3.yandex.net/datasets/data.csv'
df = pd.read_csv(url)

In [133]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [134]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [135]:
print('Missing values:\n',df.isna().sum())

Missing values:
 children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


In [136]:
print('The proportion of missing values:\n',df.isna().sum()/df.isna().count())

The proportion of missing values:
 children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64


Anomalies and data errors:

In [137]:
print('Incorrect value in the "gender" column:\n',df['gender'].value_counts())

Incorrect value in the "gender" column:
 F      14236
M       7288
XNA        1
Name: gender, dtype: int64


In [138]:
print('Incorrect value in the "children" column:\n',df['children'].value_counts())

Incorrect value in the "children" column:
  0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


In [139]:
print('No values in the "dob_years" column:\n',df['dob_years'].value_counts())

No values in the "dob_years" column:
 35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


In [140]:
print('Negative and non-integer values in the "days_employed" column:\n',df['days_employed'].value_counts())

Negative and non-integer values in the "days_employed" column:
 -8437.673028      1
-3507.818775      1
 354500.415854    1
-769.717438       1
-3963.590317      1
                 ..
-1099.957609      1
-209.984794       1
 398099.392433    1
-1271.038880      1
-1984.507589      1
Name: days_employed, Length: 19351, dtype: int64


In [141]:
print('Different capitalization of the data in the "family_status" column:\n',df['family_status'].value_counts())

Different capitalization of the data in the "family_status" column:
 женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64


In [142]:
print('Different capitalization of the data in the "education" column:\n',df['education'].value_counts())

Different capitalization of the data in the "education" column:
 среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64


In [143]:
print('The data in the "purpose" column require categorization:\n',df['purpose'].value_counts())

The data in the "purpose" column require categorization:
 свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего авто

### Conclusions:

The total number of values is 21525. 

Missing values were found in the "days_employed" and "total_income" columns in the amount of 2174 or 10.1% of the total number of values for each column.

The following data anomalies have been detected:
- Incorrect value in the "gender" column;
- Incorrect values in the "children" column;
- Zero values in the "dob_years" column;
- Negative and non-integer values in the "days_employed" column;
- Different data case in the "family_status" and "education" columns;
- Data in the "purpose" column requires categorization.

These errors may be caused by:
- Errors in data extraction from the database;
- Calculation errors;
- Errors when entering information (human factor).

Possible cause of this could be data upload errors from the database, as well as data entry errors (human factor). 

Replacing missing values with median values is more objective when analyzing this dataframe, as some data stands out from the majority and using mean values would be incorrect.

## Preprocessing

### Correcting anomalies, errors, and omissions in data.

Let's write a function for categorizing customers by age.

In [144]:

def days_employed(row):
    """
    Function for categorizing customers by age
    """
    age = row['dob_years']
    if age <= 30:
        return 'Group 1' 
    if age <= 40 and age > 30:
        return 'Group 2'
    if age <= 55 and age > 40:
        return 'Group 3'
    if age <= 80 and age > 55:
        return 'Group 4'

df['age_group'] = df.apply(days_employed, axis=1)

df['days_employed'] = df['days_employed'].fillna(0)
df.isna().sum()

df.groupby('age_group')['days_employed'].mean()[0]

df.loc[(df['age_group'] == '1 группа') & (df['days_employed'] == 0 ), 'days_employed'] = df.groupby('age_group')['days_employed'].mean()[0]
df.loc[(df['age_group'] == '2 группа') & (df['days_employed'] == 0 ), 'days_employed'] = df.groupby('age_group')['days_employed'].mean()[1]
df.loc[(df['age_group'] == '3 группа') & (df['days_employed'] == 0 ), 'days_employed'] = df.groupby('age_group')['days_employed'].mean()[2]
df.loc[(df['age_group'] == '4 группа') & (df['days_employed'] == 0 ), 'days_employed'] = df.groupby('age_group')['days_employed'].mean()[3]

Filling in the missing values in the total_income column (monthly income).

In [145]:
df['total_income'] = df['total_income'].fillna(0)

df.groupby('income_type')['total_income'].median()

df.loc[(df['total_income'] == 0) & (df['income_type'] == 'безработный'), 'total_income'] = df.groupby('income_type')['total_income'].median()[0]
df.loc[(df['total_income'] == 0) & (df['income_type'] == 'в декрете'), 'total_income'] = df.groupby('income_type')['total_income'].median()[1]
df.loc[(df['total_income'] == 0) & (df['income_type'] == 'госслужащий'), 'total_income'] = df.groupby('income_type')['total_income'].median()[2]
df.loc[(df['total_income'] == 0) & (df['income_type'] == 'компаньон'), 'total_income'] = df.groupby('income_type')['total_income'].median()[3]
df.loc[(df['total_income'] == 0) & (df['income_type'] == 'пенсионер'), 'total_income'] = df.groupby('income_type')['total_income'].median()[4]
df.loc[(df['total_income'] == 0) & (df['income_type'] == 'предприниматель'), 'total_income'] = df.groupby('income_type')['total_income'].median()[5]
df.loc[(df['total_income'] == 0) & (df['income_type'] == 'сотрудник'), 'total_income'] = df.groupby('income_type')['total_income'].median()[6]
df.loc[(df['total_income'] == 0) & (df['income_type'] == 'студент'), 'total_income'] = df.groupby('income_type')['total_income'].median()[7]
#print('Пропущенные значения:\n',df.isna().sum())


Correct the incorrect values in the children column.

In [146]:
df['children'] = df['children'].replace(20, 2)
df['children'] = df['children'].replace(-1, 1)
#print(df['children'].value_counts())

Deleting one row with a missing value in the gender column.

In [147]:
df = df[df['gender'] != 'XNA']

Replace the zero values in the dob_years column with the median values.

In [148]:
income_type_gr = df.groupby('income_type')['dob_years'].median()
#print(income_type_gr) 

df.loc[(df['income_type'] == 'пенсионер') & (df['dob_years'] == 0), 'dob_years'] = 60
df.loc[(df['income_type'] == 'студент') & (df['dob_years'] == 0), 'dob_years'] = 22
df.loc[(df['income_type'] == 'сотрудник') & (df['dob_years'] == 0), 'dob_years'] = 39
df.loc[(df['income_type'] == 'предприниматель') & (df['dob_years'] == 0), 'dob_years'] = 42.5
df.loc[(df['income_type'] == 'компаньон') & (df['dob_years'] == 0), 'dob_years'] = 39
df.loc[(df['income_type'] == 'госслужащий') & (df['dob_years'] == 0), 'dob_years'] = 40
df.loc[(df['income_type'] == 'в декрете') & (df['dob_years'] == 0), 'dob_years'] = 39
df.loc[(df['income_type'] == 'безработный ') & (df['dob_years'] == 0), 'dob_years'] = 38
#print(df['dob_years'].isnull().sum())

/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


Correcting errors (negative values) in the days_employed column.

In [149]:
df['days_employed'] = abs(df['days_employed'])

In [150]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21524 non-null  int64  
 1   days_employed     21524 non-null  float64
 2   dob_years         21524 non-null  float64
 3   education         21524 non-null  object 
 4   education_id      21524 non-null  int64  
 5   family_status     21524 non-null  object 
 6   family_status_id  21524 non-null  int64  
 7   gender            21524 non-null  object 
 8   income_type       21524 non-null  object 
 9   debt              21524 non-null  int64  
 10  total_income      21524 non-null  float64
 11  purpose           21524 non-null  object 
 12  age_group         21524 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.3+ MB


### Change of data types.

In [151]:
df['total_income'] = df['total_income'].astype('int')
df['dob_years'] = df['dob_years'].astype('int')
df['days_employed'] = df['days_employed'].astype('int')

### Removal of duplicates.

In [152]:
df.drop_duplicates().reset_index()

index  children  days_employed  dob_years education  education_id  \
0          0         1           8437         42    высшее             0   
1          1         1           4024         36   среднее             1   
2          2         0           5623         33   Среднее             1   
3          3         3           4124         32   среднее             1   
4          4         0         340266         53   среднее             1   
...      ...       ...            ...        ...       ...           ...   
21465  21520         1           4529         43   среднее             1   
21466  21521         0         343937         67   среднее             1   
21467  21522         1           2113         38   среднее             1   
21468  21523         3           3112         38   среднее             1   
21469  21524         2           1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21465  гражданский брак                 1      F   компаньон     0   
21466   женат / замужем                 0      F   пенсионер     0   
21467  гражданский брак                 1      M   сотрудник     1   
21468   женат / замужем                 0      M   сотрудник     1   
21469   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose age_group  
0            253875                 покупка жилья   Group 3  
1            112080       приобретение автомобиля   Group 2  
2            145885                 покупка жилья   Group 2  
3            267628    дополнительное образование   Group 2  
4            158616               сыграть свадьбу   Group 3  
...             ...                           ...       ...  
21465        224791             операции с жильем   Group 3  
21466        155999          сделка с автомобилем   Group 4  
21467         89672                  недвижимость   Group 2  
21468        244093  на покупку своего автомобиля   Group 2  
21469         82047         на покупку автомобиля   Group 2  

[21470 rows x 14 columns]

### Additionally

Standardizing the data in the "education" and "family_status" columns to one case.

In [153]:
df['education'] = df['education'].str.lower()
df['family_status'] = df['family_status'].str.lower()

Categorizing the data in the "purpose" column (the purpose of obtaining the credit).

In [154]:
def rename_purpose(row):
    """
    Function for categorizing by credit purpose
    """
    purpose = row['purpose']
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'недвиж' in purpose:
        return 'операции с недвижимостью'
    if 'свадь' in purpose:
        return 'проведение свадьбы'
    return 'получение образования'
df['purpose_category'] = df.apply(rename_purpose, axis=1)

#print(df['purpose'].value_counts())
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1           8437         42               высшее             0   
1          1           4024         36              среднее             1   
2          0           5623         33              среднее             1   
3          3           4124         32              среднее             1   
4          0         340266         53              среднее             1   
5          0            926         27               высшее             0   
6          0           2879         43               высшее             0   
7          0            152         50              среднее             1   
8          2           6929         35               высшее             0   
9          0           2188         41              среднее             1   
10         2           4171         36               высшее             0   
11         0            792         40              среднее             1   
12         0              0         65              среднее             1   
13         0           1846         54  неоконченное высшее             2   
14         0           1844         56               высшее             0   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        110179   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   

                              purpose age_group          purpose_category  
0                       покупка жилья   Group 3  операции с недвижимостью  
1             приобретение автомобиля   Group 2    операции с автомобилем  
2                       покупка жилья   Group 2  операции с недвижимостью  
3          дополнительное образование   Group 2     получение образования  
4                     сыграть свадьбу   Group 3        проведение свадьбы  
5                       покупка жилья   Group 1  операции с недвижимостью  
6                   операции с жильем   Group 3  операции с недвижимостью  
7                         образование   Group 3     получение образования  
8               на проведение свадьбы   Group 2        проведение свадьбы  
9             покупка жилья для семьи   Group 3  операции с недвижимостью  
10               покупка недвижимости   Group 2  операции с недвижимостью  
11  покупка коммерческой недвижимости   Group 2  операции с недвижимостью  
12                    сыграть свадьбу   Group 4        проведение свадьбы  
13            приобретение автомобиля   Group 3    операции с автомобилем  
14         покупка жилой недвижимости   Group 4  операции с недвижимостью

A manual search for duplicates with case sensitivity was applied to the dataframe, as not all columns had duplicates. 

The data in the education (client's level of education) and family_status (family status) columns were brought to a single case.  The data in the "purpose" column (the purpose of obtaining the credit) is categorized and the apply() function was used for data categorization.

Possible cause of the duplicate data is related to the human factor when manually entering data at the time of customer service.

### Forming additional dataframes of dictionaries, decomposing the original dataframe.

In [155]:
education_dict = df[['education', 'education_id']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
family_status_dict = df[['family_status', 'family_status_id']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
display(education_dict)
display(family_status_dict)

#df = df.drop(['education','family_status'],axis=1)
#display(df.head(15))

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

### Categorization of income.

In [156]:
def total_income_category (row):
    """
    Function for categorizing customers by income
    """
    total_income_group=row['total_income']
    if total_income_group<=30000:
        return 'E'
    if total_income_group<=50000:
        return 'D'
    if total_income_group<=200000:
        return 'C'
    if total_income_group<=1000000:
        return 'B'
    return 'A'
df['total_income_category'] = df.apply(total_income_category, axis=1)
display(df.head(15))

children  days_employed  dob_years            education  education_id  \
0          1           8437         42               высшее             0   
1          1           4024         36              среднее             1   
2          0           5623         33              среднее             1   
3          3           4124         32              среднее             1   
4          0         340266         53              среднее             1   
5          0            926         27               высшее             0   
6          0           2879         43               высшее             0   
7          0            152         50              среднее             1   
8          2           6929         35               высшее             0   
9          0           2188         41              среднее             1   
10         2           4171         36               высшее             0   
11         0            792         40              среднее             1   
12         0              0         65              среднее             1   
13         0           1846         54  неоконченное высшее             2   
14         0           1844         56               высшее             0   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        110179   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   

                              purpose age_group          purpose_category  \
0                       покупка жилья   Group 3  операции с недвижимостью   
1             приобретение автомобиля   Group 2    операции с автомобилем   
2                       покупка жилья   Group 2  операции с недвижимостью   
3          дополнительное образование   Group 2     получение образования   
4                     сыграть свадьбу   Group 3        проведение свадьбы   
5                       покупка жилья   Group 1  операции с недвижимостью   
6                   операции с жильем   Group 3  операции с недвижимостью   
7                         образование   Group 3     получение образования   
8               на проведение свадьбы   Group 2        проведение свадьбы   
9             покупка жилья для семьи   Group 3  операции с недвижимостью   
10               покупка недвижимости   Group 2  операции с недвижимостью   
11  покупка коммерческой недвижимости   Group 2  операции с недвижимостью   
12                    сыграть свадьбу   Group 4        проведение свадьбы   
13            приобретение автомобиля   Group 3    операции с автомобилем   
14         покупка жилой недвижимости   Group 4  операции с недвижимостью   

   total_income_category  
0                      B  
1                      C  
2                      C  
3                      B  
4                      C  
5                      B  
6                      B  
7                      C 

### Categorization of loan objectives.

In [157]:
def rename_purpose(row):
    """
    Function for categorizing customers by loan purpose
    """
    purpose = row['purpose']
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'недвиж' in purpose:
        return 'операции с недвижимостью'
    if 'свадь' in purpose:
        return 'проведение свадьбы'
    return 'получение образования'
df['purpose_category'] = df.apply(rename_purpose, axis=1)
display(df.head(15))

children  days_employed  dob_years            education  education_id  \
0          1           8437         42               высшее             0   
1          1           4024         36              среднее             1   
2          0           5623         33              среднее             1   
3          3           4124         32              среднее             1   
4          0         340266         53              среднее             1   
5          0            926         27               высшее             0   
6          0           2879         43               высшее             0   
7          0            152         50              среднее             1   
8          2           6929         35               высшее             0   
9          0           2188         41              среднее             1   
10         2           4171         36               высшее             0   
11         0            792         40              среднее             1   
12         0              0         65              среднее             1   
13         0           1846         54  неоконченное высшее             2   
14         0           1844         56               высшее             0   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        110179   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   

                              purpose age_group          purpose_category  \
0                       покупка жилья   Group 3  операции с недвижимостью   
1             приобретение автомобиля   Group 2    операции с автомобилем   
2                       покупка жилья   Group 2  операции с недвижимостью   
3          дополнительное образование   Group 2     получение образования   
4                     сыграть свадьбу   Group 3        проведение свадьбы   
5                       покупка жилья   Group 1  операции с недвижимостью   
6                   операции с жильем   Group 3  операции с недвижимостью   
7                         образование   Group 3     получение образования   
8               на проведение свадьбы   Group 2        проведение свадьбы   
9             покупка жилья для семьи   Group 3  операции с недвижимостью   
10               покупка недвижимости   Group 2  операции с недвижимостью   
11  покупка коммерческой недвижимости   Group 2  операции с недвижимостью   
12                    сыграть свадьбу   Group 4        проведение свадьбы   
13            приобретение автомобиля   Group 3    операции с автомобилем   
14         покупка жилой недвижимости   Group 4  операции с недвижимостью   

   total_income_category  
0                      B  
1                      C  
2                      C  
3                      B  
4                      C  
5                      B  
6                      B  
7                      C 

### Conclusions:

As a result of the preparation conducted, the data frame in "Data Preprocessing" had the errors, anomalies, missing values, and duplicates identified in "Data Overview" - corrected.

Based on the character of the identified defects provided for data analysis, I recommend that the customer pay attention to the quality of customer data collection, especially the data that the operator enters manually. I also pay attention to the quality of the data obtained by calculation, especially in the "days_employed" column, the initial state of which does not allow for further comprehensive analysis.

## Hypothesis testing.

#### Hypothesis 1: Is there a correlation between the number of children and the repayment of the loan on time?

In [158]:
data_pivot_children = df.pivot_table(index=['debt'], columns='children', values = 'family_status_id', aggfunc='mean')
display(data_pivot_children)

print('The proportion of customers with debts with a certain number of children to customers with the same number of children but without debts \n(higher % - lower return of credit in time) is higher:')

print('{0:.2f}% Нет детей'.format((data_pivot_children[0][1] / data_pivot_children[0][0])))
print('{0:.2f}% 1 - ребенок'.format((data_pivot_children[1][1] / data_pivot_children[1][0])))
print('{0:.2f}% 2 - ребенка'.format((data_pivot_children[2][1] / data_pivot_children[2][0])))
print('{0:.2f}% 3 - ребенка'.format((data_pivot_children[3][1] / data_pivot_children[3][0])))
print('{0:.2f}% 4 - ребенка'.format((data_pivot_children[4][1] / data_pivot_children[4][0])))
print('{0:.2f}% 5 - детей'.format((data_pivot_children[5][1] / data_pivot_children[5][0])))

children         0         1         2         3         4         5
debt                                                                
0         1.109591  0.797738  0.459305  0.389439  0.459459  0.222222
1         1.260583  0.905618  0.509901  0.555556  1.000000       NaN

The proportion of customers with debts with a certain number of children to customers with the same number of children but without debts 
(higher % - lower return of credit in time) is higher:
1.14% Нет детей
1.14% 1 - ребенок
1.11% 2 - ребенка
1.43% 3 - ребенка
2.18% 4 - ребенка
nan% 5 - детей


#### Conclusions: On average, having a child increases the likelihood of missing a loan payment.

#### Hypothesis 2: Is there a correlation between marital status and loan repayment on time?

In [159]:
data_pivot_family_status = df.pivot_table(index=['debt'], columns='family_status', values='age_group', aggfunc='count')
display(data_pivot_family_status)

print('The proportion of customers with debts in a certain family status to customers in the same family status but without debts \n(above % - lower return of credit in time):')

print('{0:.2f}% не женат / не замужем'.format((data_pivot_family_status['не женат / не замужем'][1] / data_pivot_family_status['не женат / не замужем'][0])*100))
print('{0:.2f}% в разводе'.format((data_pivot_family_status['в разводе'][1] / data_pivot_family_status['в разводе'][0])*100))
print('{0:.2f}% вдовец / вдова'.format((data_pivot_family_status['вдовец / вдова'][1] / data_pivot_family_status['вдовец / вдова'][0])*100))
print('{0:.2f}% гражданский брак'.format((data_pivot_family_status['гражданский брак'][1] / data_pivot_family_status['гражданский брак'][0])*100))
print('{0:.2f}% женат / замужем'.format((data_pivot_family_status['женат / замужем'][1] / data_pivot_family_status['женат / замужем'][0])*100))

family_status  в разводе  вдовец / вдова  гражданский брак  женат / замужем  \
debt                                                                          
0                   1110             897              3788            11449   
1                     85              63               388              931   

family_status  не женат / не замужем  
debt                                  
0                               2539  
1                                274

The proportion of customers with debts in a certain family status to customers in the same family status but without debts 
(above % - lower return of credit in time):
10.79% не женат / не замужем
7.66% в разводе
7.02% вдовец / вдова
10.24% гражданский брак
8.13% женат / замужем


не женат / не замужем -  unmarried

в разводе - divorced

вдовец / вдова - mwidower / widow

гражданский брак - civil marriage

женат / замужем - married

#### Conclusions: Customers with a marital status of "not married" or "civil union" are less likely to pay off their credit on time.

#### Hypothesis 3: Is there a correlation between income level and loan repayment on time?

In [160]:
data_pivot_income_gr = df.pivot_table(index=['debt'], columns='total_income_category', values='age_group', aggfunc='count')
display(data_pivot_income_gr)

print('The proportion of customers with debts to customers with the same income level but without debts \n(higher % - lower return on credit in time) is lower:')
print('{0:.2f}% Group A with income 1000001 and higher'.format((data_pivot_income_gr['A'][1] / data_pivot_income_gr['A'][0])*100))
print('{0:.2f}% Group B with income 200001–1000000'.format((data_pivot_income_gr['B'][1] / data_pivot_income_gr['B'][0])*100))
print('{0:.2f}% Group C with income 50001–200000'.format((data_pivot_income_gr['C'][1] / data_pivot_income_gr['C'][0])*100))
print('{0:.2f}% Group D with income 30001–50000'.format((data_pivot_income_gr['D'][1] / data_pivot_income_gr['D'][0])*100))
print('{0:.2f}% Group E with income up to 30000'.format((data_pivot_income_gr['E'][1] / data_pivot_income_gr['E'][0])*100))

total_income_category   A     B      C    D   E
debt                                           
0                      23  4685  14726  329  20
1                       2   356   1360   21   2

The proportion of customers with debts to customers with the same income level but without debts 
(higher % - lower return on credit in time) is lower:
8.70% Group A with income 1000001 and higher
7.60% Group B with income 200001–1000000
9.24% Group C with income 50001–200000
6.38% Group D with income 30001–50000
10.00% Group E with income up to 30000


#### Conclusions: Customers with incomes of "30,001 - 50,000" and "200,001 - 1,000,000" usually pay off their credit on time.

#### Hypothesis 4: How do different loan purposes affect its repayment on time?

In [161]:
data_pivot_purpose = df.pivot_table(index=['debt'], columns='purpose_category', values='age_group', aggfunc='count')
display(data_pivot_purpose)

print('The proportion of customers with debts for a certain purpose of credit to customers with the same purpose but without debts \n(higher % - lower return of credit in time):')
print('{0:.2f}% операции с автомобилем'.format((data_pivot_purpose['операции с автомобилем'][1] / data_pivot_purpose['операции с автомобилем'][0])*100))
print('{0:.2f}% операции с недвижимостью'.format((data_pivot_purpose['операции с недвижимостью'][1] / data_pivot_purpose['операции с недвижимостью'][0])*100))
print('{0:.2f}% получение образования'.format((data_pivot_purpose['получение образования'][1] / data_pivot_purpose['получение образования'][0])*100))
print('{0:.2f}% проведение свадьбы'.format((data_pivot_purpose['проведение свадьбы'][1] / data_pivot_purpose['проведение свадьбы'][0])*100))

purpose_category  операции с автомобилем  операции с недвижимостью  \
debt                                                                 
0                                   3912                     10057   
1                                    403                       782   

purpose_category  получение образования  проведение свадьбы  
debt                                                         
0                                  3652                2162  
1                                   370                 186

The proportion of customers with debts for a certain purpose of credit to customers with the same purpose but without debts 
(higher % - lower return of credit in time):
10.30% операции с автомобилем
7.78% операции с недвижимостью
10.13% получение образования
8.60% проведение свадьбы


операции с автомобилем - operations with a car

операции с недвижимостью - operations with real estate

получение образования - obtaining an education

проведение свадьбы - holding a wedding

#### Conclusions: Customers more often pay off loans taken for weddings and real estate operations on time.

## Overall conclusion:

When working with the data, missing values were filled in, duplicates were removed, and categories were identified for loan purposes, age groups, education, and marital status. Using pivot tables, an assessment of the criteria affecting loan repayment was made.

According to the research and hypothesis testing, the following dependencies have been identified:
- When the number of children of the client increases, the return of the loan on time decreases;
- Customers with family status "not married" and "civil marriage" less often pay off the loan on time;
- Customers with income "30 001 - 50 000" and "200 001 - 1 000 000" most often pay off the loan on time;
- Credits taken for weddings and real estate operations are usually paid on time..

To avoid the risk of non-repayment of loans, the Customer should pay attention to the Complaints service and Underwriters specialists for the above-mentioned categories of customers when making a decision to issue a loan. 

Additional analysis can help more accurately predict the probability of loan repayment on time, as well as show the relationship between the hypotheses being tested.